File: apps/host/internal/endpoints/endpoints_test.go
package endpoints

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
)

func TestPublishKeepsTheOtherServiceAndStaysPrivate(t *testing.T) {
	path := Path(t.TempDir())
	runtimeRecord := Now("runtime-instance")
	runtimeRecord.HTTP = "http://127.0.0.1:53211"
	runtimeRecord.Socket = "/tmp/armadra/runtime.sock"
	if err := Publish(path, RuntimeService, runtimeRecord); err != nil {
		t.Fatalf("publish runtime: %v", err)
	}
	hostRecord := Now("host-instance")
	hostRecord.HTTP = "http://127.0.0.1:53212"
	if err := Publish(path, HostService, hostRecord); err != nil {
		t.Fatalf("publish host: %v", err)
	}

	document := Read(path)
	if document.Version != Version {
		t.Fatalf("version = %d, want %d", document.Version, Version)
	}
	if document.Runtime == nil || document.Runtime.HTTP != "http://127.0.0.1:53211" {
		t.Fatalf("the Runtime record did not survive the Host publish: %+v", document.Runtime)
	}
	if document.Host == nil || document.Host.HTTP != "http://127.0.0.1:53212" {
		t.Fatalf("host record = %+v", document.Host)
	}
	if document.Host.ProcessID != uint32(os.Getpid()) {
		t.Fatalf("processId = %d, want this process", document.Host.ProcessID)
	}

	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read: %v", err)
	}
	// The Rust Runtime reads this file too; the keys are its camelCase.
	for _, key := range []string{`"instanceId"`, `"writtenAt"`, `"processId"`} {
		if !strings.Contains(string(raw), key) {
			t.Fatalf("%s missing from %s", key, raw)
		}
	}
	// A record with no WebSocket must not claim an empty one.
	if strings.Contains(string(raw), `"websocket"`) {
		t.Fatalf("absent transports should be omitted: %s", raw)
	}
	if runtime.GOOS != "windows" {
		info, err := os.Stat(path)
		if err != nil {
			t.Fatalf("stat: %v", err)
		}
		if info.Mode().Perm() != 0600 {
			t.Fatalf("mode = %v, want 0600", info.Mode().Perm())
		}
	}
}

func TestWithdrawRemovesOnlyItsOwnRecord(t *testing.T) {
	path := Path(t.TempDir())
	if err := Publish(path, RuntimeService, Now("runtime")); err != nil {
		t.Fatalf("publish runtime: %v", err)
	}
	if err := Publish(path, HostService, Now("host")); err != nil {
		t.Fatalf("publish host: %v", err)
	}
	if err := Withdraw(path, HostService); err != nil {
		t.Fatalf("withdraw: %v", err)
	}
	document := Read(path)
	if document.Host != nil {
		t.Fatalf("host record survived withdrawal: %+v", document.Host)
	}
	if document.Runtime == nil {
		t.Fatal("withdrawing the Host removed the Runtime record")
	}
	// A file that was never written is not an error to withdraw from.
	if err := Withdraw(Path(filepath.Join(t.TempDir(), "absent")), HostService); err != nil {
		t.Fatalf("withdraw from a missing file: %v", err)
	}
}

func TestACorruptOrNewerFileIsIgnoredRatherThanFatal(t *testing.T) {
	path := Path(t.TempDir())
	for _, body := range []string{"{ not json", `{"version":99,"host":{"http":"x"}}`} {
		if err := os.WriteFile(path, []byte(body), 0600); err != nil {
			t.Fatalf("seed: %v", err)
		}
		if document := Read(path); document.Host != nil || document.Runtime != nil {
			t.Fatalf("%q should have read as empty, got %+v", body, document)
		}
		if err := Publish(path, HostService, Now("host")); err != nil {
			t.Fatalf("publish over %q: %v", body, err)
		}
		if Read(path).Host == nil {
			t.Fatalf("publish over %q did not take", body)
		}
	}
}

func TestAnUnknownServiceIsRefusedRatherThanSilentlyDropped(t *testing.T) {
	path := Path(t.TempDir())
	if err := Publish(path, "worker", Now("worker")); err == nil {
		t.Fatal("an unknown service key should be an error")
	}
	if _, err := os.Stat(path); err == nil {
		t.Fatal("a refused publish must not create the file")
	}
}

// The Rust Runtime and this package have to agree on the wire shape, because
// they write the same file. This pins the field names that agreement rests on.
func TestTheDocumentShapeMatchesTheRuntimeContract(t *testing.T) {
	record := Now("instance")
	record.HTTP = "http://127.0.0.1:1"
	record.WebSocket = "ws://127.0.0.1:1"
	record.Socket = "/tmp/s.sock"
	record.Pipe = `\\.\pipe\armadra`
	body, err := json.Marshal(Document{Version: Version, Host: &record})
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	var decoded map[string]any
	if err := json.Unmarshal(body, &decoded); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	host, ok := decoded["host"].(map[string]any)
	if !ok {
		t.Fatalf("host is not an object: %s", body)
	}
	for _, key := range []string{"instanceId", "writtenAt", "processId", "http", "websocket", "socket", "pipe"} {
		if _, present := host[key]; !present {
			t.Fatalf("%q missing from %s", key, body)
		}
	}
}

File: apps/desktop/src/core/endpoints.test.ts
import { readFileSync, statSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { describe, expect, it } from "vitest";
import {
  ENDPOINTS_VERSION,
  HOST_SERVICE,
  RUNTIME_SERVICE,
  publish,
  read,
  serviceEndpointNow,
  withdraw,
} from "./endpoints";
import { endpointsFile } from "./paths";
import { tempDir } from "./testing/temp-dir";

function temporary(): string {
  return endpointsFile(tempDir("armadra-endpoints-"));
}

const runtimeRecord = () => ({
  ...serviceEndpointNow("runtime-instance"),
  http: "http://127.0.0.1:53211",
  socket: "/tmp/armadra/runtime.sock",
});

describe("endpoints.json", () => {
  it("stamps a record with this process and an RFC 3339 time", () => {
    const record = serviceEndpointNow("abc", () => new Date(0), 4321);
    expect(record).toEqual({
      instanceId: "abc",
      writtenAt: "1970-01-01T00:00:00.000Z",
      processId: 4321,
    });
    expect(serviceEndpointNow("abc").processId).toBe(process.pid);
  });

  it("keeps the other service and stays camelCase", () => {
    const path = temporary();
    publish(path, RUNTIME_SERVICE, runtimeRecord());
    publish(path, HOST_SERVICE, {
      ...serviceEndpointNow("host-instance"),
      http: "http://127.0.0.1:53212",
    });

    const raw = readFileSync(path, "utf8");
    expect(raw).toContain('"instanceId"');
    expect(raw).toContain('"writtenAt"');
    expect(raw).toContain('"processId"');
    // Absent transports are omitted rather than written as null.
    expect(raw).not.toContain("null");
    expect(raw.endsWith("\n")).toBe(true);

    const document = read(path);
    expect(document.version).toBe(ENDPOINTS_VERSION);
    expect(document.runtime?.http).toBe("http://127.0.0.1:53211");
    expect(document.host?.http).toBe("http://127.0.0.1:53212");
    expect(document.runtime?.processId).toBe(process.pid);
  });

  it("writes the keys in the order the Rust struct declares them", () => {
    const path = temporary();
    publish(path, RUNTIME_SERVICE, runtimeRecord());
    const parsed = JSON.parse(readFileSync(path, "utf8")) as Record<
      string,
      unknown
    >;
    expect(Object.keys(parsed)).toEqual(["version", "runtime"]);
    expect(Object.keys(parsed.runtime as object)).toEqual([
      "instanceId",
      "writtenAt",
      "processId",
      "http",
      "socket",
    ]);
  });

  it("withdraws one service without removing the file or the other record", () => {
    const path = temporary();
    publish(path, RUNTIME_SERVICE, runtimeRecord());
    publish(path, HOST_SERVICE, serviceEndpointNow("host"));
    withdraw(path, RUNTIME_SERVICE);
    const document = read(path);
    expect(document.runtime).toBeUndefined();
    expect(document.host).toBeDefined();
  });

  it("does not mind withdrawing from a file that was never written", () => {
    expect(() =>
      withdraw(join(tempDir("armadra-"), "gone.json"), HOST_SERVICE),
    ).not.toThrow();
  });

  it("reads a corrupt or newer document as empty and rewrites it", () => {
    const path = temporary();
    writeFileSync(path, "{ not json");
    expect(read(path)).toEqual({ version: ENDPOINTS_VERSION });
    writeFileSync(path, '{"version":99,"runtime":{}}');
    expect(read(path)).toEqual({ version: ENDPOINTS_VERSION });
    writeFileSync(path, "[]");
    expect(read(path)).toEqual({ version: ENDPOINTS_VERSION });
    publish(path, RUNTIME_SERVICE, runtimeRecord());
    expect(read(path).runtime).toBeDefined();
  });

  it("reads a missing file as empty rather than failing a start-up", () => {
    expect(
      read(join(tmpdir(), "armadra-does-not-exist", "endpoints.json")),
    ).toEqual({
      version: ENDPOINTS_VERSION,
    });
  });

  it("tolerates a record whose fields are the wrong type", () => {
    const path = temporary();
    writeFileSync(
      path,
      JSON.stringify({
        version: 1,
        runtime: { instanceId: 7, processId: "x" },
      }),
    );
    expect(read(path).runtime).toEqual({
      instanceId: "",
      writtenAt: "",
      processId: 0,
      http: undefined,
      websocket: undefined,
      socket: undefined,
      pipe: undefined,
    });
  });

  it.skipIf(process.platform === "win32")(
    "writes the file 0600 in a 0700 directory",
    () => {
      const path = temporary();
      publish(path, RUNTIME_SERVICE, runtimeRecord());
      expect(statSync(path).mode & 0o777).toBe(0o600);
      expect(statSync(join(path, "..")).mode & 0o777).toBe(0o700);
    },
  );
});
